pkg/genyaml/genyaml.go

```diff
@@ -0,0 +1,404 @@
+	// number type, so we can preserve number type throughout this process.
+	err := yaml3.Unmarshal(j, &jsonObj)
+	if err != nil {
+		return nil, err
+	}
+
+	// marshal this object into YAML.
+	return yaml3.Marshal(jsonObj)
+}
+
+// astFrom takes a path to a Go file and returns the abstract syntax tree (AST) for that file.
+func astFrom(path string) (*doc.Package, error) {
+	fset := token.NewFileSet()
+	m := make(map[string]*ast.File)
+
+	f, err := parser.ParseFile(fset, path, nil, parser.ParseComments)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse file to AST from path: %s", path)
+	}
+
+	m[path] = f
+	apkg, _ := ast.NewPackage(fset, m, nil, nil)
+
+	astDoc := doc.New(apkg, "", 0)
+	if astDoc == nil {
+		return nil, fmt.Errorf("unable to parse AST documentation from path: %s", path)
+	}
+
+	return astDoc, nil
+}
+
+// fmtRawDoc formats/sanitizes a Go doc string removing TODOs, newlines, whitespace, and various other characters from the resultant string.
+func fmtRawDoc(rawDoc string) string {
+	var buffer bytes.Buffer
+	delPrevChar := func() {
+		if buffer.Len() > 0 {
+			buffer.Truncate(buffer.Len() - 1) // Delete the last " " or "\n".
+		}
+	}
+
+	// Ignore all lines after ---.
+	rawDoc = strings.Split(rawDoc, "---")[0]
+
+	for _, line := range strings.Split(rawDoc, "\n") {
+		line = strings.TrimRight(line, " ")
+		leading := strings.TrimLeft(line, " ")
+		switch {
+		case len(line) == 0: // Keep paragraphs
+			delPrevChar()
+			buffer.WriteString("\n\n")
+		case strings.HasPrefix(leading, "TODO"): // Ignore one line TODOs.
+		case strings.HasPrefix(leading, "+"): // Ignore instructions to the generators.
+		default:
+			line += " "
+			buffer.WriteString(line)
+		}
+	}
+
+	postDoc := strings.TrimRight(buffer.String(), "\n")
+	postDoc = strings.Trim(postDoc, " ")
+	postDoc = strings.Replace(postDoc, "\\\"", "\"", -1) // replace user's \" to ".
+	postDoc = strings.Replace(postDoc, "\"", "\\\"", -1) // Escape ".
+	postDoc = strings.Replace(postDoc, "\n", "\\n", -1)
+	postDoc = strings.Replace(postDoc, "\t", "\\t", -1)
+
+	return postDoc
+}
+
+// fieldName extracts the name of the field as it should appear in YAML format and returns the resultant string.
+// "-" indicates that this field is not part of the YAML representation and is thus excluded.
+func fieldName(field *ast.Field, tag string) string {
+	tagVal := ""
+	if field.Tag != nil {
+		tagVal = reflect.StructTag(field.Tag.Value[1 : len(field.Tag.Value)-1]).Get(tag) // Delete first and last quotation.
+		if strings.Contains(tagVal, "inline") {
+			return "-"
+		}
+	}
+
+	tagVal = strings.Split(tagVal, ",")[0] // This can return "-".
+	if tagVal == "" {
+		return fieldType(field)
+	}
+	return tagVal
+}
+
+// fieldType extracts the type of the field and returns the resultant string.
+func fieldType(field *ast.Field) string {
+	if field.Names != nil {
+		return field.Names[0].Name
+	}
+	return field.Type.(*ast.Ident).Name
```

Comment:
When I tried this out, I got a panic here:

```
panic: interface conversion: ast.Expr is *ast.SelectorExpr, not *ast.Ident
```

Reply:
> When I tried it out there was unfortunately a panic, see below.

🤔. I can handle the case where the type assertion fails. It would be helpful if you can direct me to the *file* you used that resulted in this error.

> Would it be possible as a follow-up to build this in a way that we can just feed it `.go` files and it will emit docs? That would be a lot nicer from an UX perspective. Afaik the various go linters have code to read and analyse go code

It is definitely possible, but I would need to understand the desired *inputs* and *outputs*. I tried to make this generic enough where the functions could be reused. Currently, this is designed specifically for #13507 where the input is a `.go` file and *initialized* `struct` and the *output* is YAML (from `struct`) combined with comments (from `.go`)

Reply:
I see. It is panicking on a `select` statement. We are only concerned about structs here so I can add the appropriate checks and skip irrelevant lines.

Reply:
I added [handling](https://github.com/kubernetes/test-infra/pull/13637/commits/1fa7ece6ca13c0aa12abf03c3e809f97163ed66c#diff-a55961ed7ceed93fe28ff3b9d47a8802R217-R240) for `SelectorExpr`, albeit these only work *properly* if they are [tagged](https://github.com/kubernetes/test-infra/pull/13637/commits/1fa7ece6ca13c0aa12abf03c3e809f97163ed66c#diff-a55961ed7ceed93fe28ff3b9d47a8802R217-R240).